src/DebugEngineHost/HostLogger.cs

```diff
@@ -3,101 +3,151 @@
+        /// </summary>
+        Debug,
+        /// <summary>
+        /// Logs that track the general flow of the application.
+        /// These logs should have long-term value.
+        /// </summary>
+        Information,
+        /// <summary>
+        /// Logs that highlight an abnormal or unexpected event in the application flow, but do not otherwise cause the application execution to stop.
+        /// </summary>
+        Warning,
+        /// <summary>
+        /// Logs that highlight when the current flow of execution is stopped due to a failure.
+        /// These should indicate a failure in the current activity, not an application-wide failure.
+        /// </summary>
+        Error,
+        /// <summary>
+        /// Logs that describe an unrecoverable application or system crash, or a catastrophic failure that requires immediate attention.
+        /// </summary>
+        Critical,
+        /// <summary>
+        /// Not used for writing log messages.
+        /// Specifies that a logging category should not write any messages.
         /// </summary>
-        /// <param name="outputString"></param>
-        public delegate void OutputCallback(string outputString);
+        None
+    }
 
-        private StreamWriter _streamWriter;
-        private OutputCallback _callback;
-        private readonly object _locker = new object();
+    public class HostLogChannel
+    {
+        private readonly Action<LogLevel, string> _log;
+        private readonly StreamWriter _logFile;
 
-        internal HostLogger(StreamWriter streamWriter = null, OutputCallback callback = null)
-        {
-            _streamWriter = streamWriter;
-            _callback = callback;
-        }
+        private readonly object _lock = new object();
 
-        public void WriteLine(string line)
+        public HostLogChannel(Action<LogLevel, string> logAction, string file)
         {
-            lock (_locker)
+            _log = logAction;
+
+            if (!string.IsNullOrEmpty(file))
             {
-                if (_streamWriter != null)
-                    _streamWriter.WriteLine(line);
-                _callback?.Invoke(line);
+                _logFile = File.CreateText(file);
             }
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="verbosity"></param>
+        /// <param name="message"></param>
+        public void WriteLine(LogLevel verbosity, string message)
+        {
+            _log?.Invoke(verbosity, message);
+            _logFile?.WriteLine(message);
+            _logFile?.Flush();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="verbosity"></param>
+        /// <param name="format"></param>
+        /// <param name="values"></param>
+        public void WriteLine(LogLevel verbosity, string format, params object[] values)
```

Comment:
>LogLevel verbosity [](http://example.com/codeflow?start=30&length=18)

I think this class needs a minimum verbosity and ignores attempts to write if it doesn't meet it.